### ISSUE TITLE
dcrd got stuck processing blocks and is spinning a cpu

### ISSUE THREAD

**Issue by jrick** [decred/dcrd#620]
Haven't investigated too far yet but during the testnet reset my node stopped processing blocks and according to prstat (like top but better for illumos) it was just sitting eating about a cpu.  According to dtrace here are the functions it was calling in about a second worth of real time:

```
[root@dcrd-test0 ~]# dtrace -p 69759 -n 'pid$target:dcrd:github.com*:entry { @counts[probefunc] = count() }'
dtrace: description 'pid$target:dcrd:github.com*:entry ' matched 3757 probes
^C

  github.com/decred/dcrd/addrmgr.IsRFC6052                       7154
  github.com/decred/dcrd/addrmgr.IsRFC3964                       7155
  github.com/decred/dcrd/addrmgr.IsRFC4380                       7155
  github.com/decred/dcrd/addrmgr.IsRFC6145                       7156
  github.com/decred/dcrd/addrmgr.IsIPv4                         71312
  github.com/decred/dcrd/addrmgr.IsRFC4193                      71312
  github.com/decred/dcrd/addrmgr.IsRFC4862                      71313
  github.com/decred/dcrd/addrmgr.IsRFC1918                      71314
  github.com/decred/dcrd/addrmgr.IsRFC3849                      71314
  github.com/decred/dcrd/addrmgr.IsRFC3927                      71314
  github.com/decred/dcrd/addrmgr.GroupKey                       71315
  github.com/decred/dcrd/addrmgr.IsRFC6598                      71315
  github.com/decred/dcrd/addrmgr.IsRFC5737                      71316
  github.com/decred/dcrd/addrmgr.IsRFC2544                      71317
  github.com/decred/dcrd/addrmgr.IsRFC4843                      71318
  github.com/decred/dcrd/addrmgr.IsValid                        71318
  github.com/decred/dcrd/addrmgr.(*AddrManager).GetAddress            71319
  github.com/decred/dcrd/addrmgr.IsRoutable                     71319
  github.com/decred/dcrd/vendor/github.com/btcsuite/btclog.(*SubsystemLogger).Tracef            71319
  github.com/decred/dcrd/addrmgr.(*KnownAddress).chance            99148
  github.com/decred/dcrd/addrmgr.ipString                      142631
  github.com/decred/dcrd/addrmgr.IsLocal                       142635
  github.com/decred/dcrd/addrmgr.NetAddressKey                 142636
  github.com/decred/dcrd/addrmgr.IsOnionCatTor                 149782
```

**Comment by davecgh** [comment]
I suspect this will be remedied when the upstream connection manager code is synced in, but it's still worth a look.

**Comment by jrick** [comment]
Began looking a bit further to get an idea of what code this is.  It looks like this loop:

https://github.com/decred/dcrd/blob/9f58376efca89bb164e684933076ad93bfa6f536/server.go#L1885-L1936

And this looks like the culprit:

https://github.com/decred/dcrd/blob/9f58376efca89bb164e684933076ad93bfa6f536/server.go#L1906

If the peer is a pending peer, then it will just continue the loop without incrementing tries ever and breaking out of the loop if it exceeds the max.

**Comment by davecgh** [comment]
Yep, that has been fixed with the new upstream connection manager.

**Comment by jrick** [comment]
Fixed by #631.
